File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.io.*;
import java.util.*;

public class Main {
    static int n, temp;
    static int[][] map;
    static boolean[][] visit;
    static int[] dx = { -1, 1, 0, 0 };
    static int[] dy = { 0, 0, -1, 1 };

    public static void dfs(int x, int y) {
        visit[x][y] = true;
        for (int i = 0; i < 4; i++) {
            int mx = x + dx[i];
            int my = y + dy[i];
            if (mx >= 0 && my >= 0 && mx < n && my < n && !visit[mx][my] && map[mx][my] == 1) {
                temp++;
                dfs(mx, my);
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder stb = new StringBuilder();
        n = Integer.parseInt(br.readLine());
        temp = 0;
        map = new int[n][n];
        visit = new boolean[n][n];

        for (int i = 0; i < n; i++) {
            String line = br.readLine();
            for (int j = 0; j < n; j++) {
                map[i][j] = line.charAt(j) - '0';
            }
        }

        ArrayList<Integer> size = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (!visit[i][j] && map[i][j] == 1) {
                    temp = 1;
                    dfs(i, j);
                    size.add(temp);
                }
            }
        }

        Collections.sort(size);

        stb.append(size.size()).append("\n");
        for (int num : size) {
            stb.append(num).append("\n");
        }

        System.out.print(stb.toString());
    }
}

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.js
let [n, ...input] = require("fs")
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "input.txt")
  .toString()
  .trim()
  .split("\n");

n = Number(n);
let map = input.map((line) => line.trim().split("").map((c) => Number(c)));
let visit = Array.from(Array(n), () => Array(n).fill(false));
let temp = 0;

let dir = [[-1, 0], [1, 0], [0, -1], [0, 1]];

const dfs = (x, y) => {
  visit[x][y] = true;

  for (let i = 0; i < 4; i++) {
    let nx = x + dir[i][0];
    let ny = y + dir[i][1];

    if (nx >= 0 && nx < n && ny >= 0 && ny < n && !visit[nx][ny] && map[nx][ny] === 1) {
      temp++;
      dfs(nx, ny);
    }
  }
};

let ans = [];

for (let i = 0; i < n; i++) {
  for (let j = 0; j < n; j++) {
    if (!visit[i][j] && map[i][j] === 1) {
      temp = 1;
      dfs(i, j);
      ans.push(temp);
    }
  }
}

ans.sort((a, b) => a - b);

console.log(ans.length);
console.log(ans.join("\n"));
